python/hail/linalg/blockmatrix.py

```diff
@@ -479,6 +505,73 @@ def filter(self, rows_to_keep, cols_to_keep):
         return BlockMatrix(self._jbm.filter(jarray(Env.jvm().long, rows_to_keep),
                                             jarray(Env.jvm().long, cols_to_keep)))
 
+    @staticmethod
+    def _pos_index(i, size, name):
+        if 0 <= i < size:
+            return i
+        elif 0 <= i + size < size:
+            return i + size
+        else:
+            raise ValueError(f'invalid {name} {i} for axis of size {size}')
+
+    @staticmethod
+    def _indices_to_keep(idx, size):
+        if isinstance(idx, int):
+            i = BlockMatrix._pos_index(idx, size, 'index')
+            return [i]
+        elif isinstance(idx, slice):
+            if idx.step is None:
+                step = 1
+            else:
+                if idx.step < 1:
+                    raise ValueError(f'slice step must be positive, found {idx.step}')
+                step = idx.step
+
+            if idx.start is None:
+                start = 0
+            else:
+                start = BlockMatrix._pos_index(idx.start, size, 'start index')
+
+            if idx.stop is None:
+                stop = size
+            elif idx.stop == 0:
+                raise ValueError(f'slice {start}:{stop}:{step} is empty')
+            else:
+                stop = BlockMatrix._pos_index(idx.stop - 1, size, 'stop index') + 1
```

Comment:
Would be a bit simpler to remove the `-1` and `+1`, change the first line of `_pos_index` to `if 0 <= i <= size`, and remove the `elif idx.stop == 0` case which is caught by the `start >= stop` case at the bottom.

Ah, I see `_pos_index` is also used in `__getitem__`. But the error message in `_pos_index` would be confusing coming from here, since the stop index will be one less than what they wrote. Maybe pass `_pos_index` a flag to determine whether size is an allowed index?